LookAndFeels/foleys_Skeuomorphic.cpp

```diff
@@ -0,0 +1,241 @@
+        {
+            auto xOffset = fgRadius * 0.12f;
+            auto yOffset = fgRadius * 0.32f;
+
+            juce::ColourGradient shadowGr = makeGradient ({ centre.x + xOffset, centre.y + yOffset },
+                                                          { centre.x, centre.y + fgRadius + yOffset }, true,
+                                                       {  { 0.0,  blackA122                 },
+                                                          { 0.72, blackA122                 },
+                                                          { 1.0,  juce::Colours::transparentBlack }  });
+            fillEllipse (g, bounds.translated (xOffset, yOffset), shadowGr);
+        }
+
+        // middle
+        {
+            g.saveState();
+            clipToEllipse (g, bounds);
+
+            juce::ColourGradient gr { { 55, 55, 66 }, 0.0f, bounds.getY(),
+                                      { 34, 35, 41 }, 0.0f, bounds.getBottom(), false };
+            fillEllipse (g, bounds, gr);
+
+            juce::ColourGradient ringGr = makeGradient (centre, { centre.x, centre.y + fgRadius }, true,
+                                                     {  { 0.0,  juce::Colours::transparentBlack },
+                                                        { 0.72, juce::Colours::transparentBlack },
+                                                        { 1.0,  blackA042                 }  });
+            fillEllipse (g, bounds, ringGr);
+
+            g.setColour (whiteA042);
+            g.drawEllipse (bounds.translated (0.0f, 1.0f), 1.0f);
+            if (!isSmall) g.drawEllipse (bounds.translated (0.0f, 2.0f), 1.0f);
+
+            g.restoreState();
+            g.setColour (blackA092);
+            g.drawEllipse (bounds, 1.0f);
+        }
+    }
+
+    knobsBgs.insert ({diameter, knobImages { std::move (bg), std::move (fg) }});
+    return knobsBgs.find (diameter)->second;
+}
+
+void Skeuomorphic::drawRotarySlider (juce::Graphics& g, int x, int y, int width, int height, float sliderPos,
+                                     const float rotaryStartAngle, const float rotaryEndAngle, juce::Slider& slider)
+{
+    int diameter = (width > height)? height : width;
+    if (diameter < 16) return;
+
+    juce::Point<float> centre (x + int (width * 0.5f + 0.5f), y + int (height * 0.5f + 0.5f));
+    diameter -= (diameter % 2)? 9 : 8;
+    float radius = diameter * 0.5f;
+    x = int (centre.x - radius);
+    y = int (centre.y - radius);
+
+    juce::Rectangle<float> bounds (x, y, diameter, diameter);
+
+    bool isTiny  = diameter < 20;
+    bool isSmall = diameter < 60;
+
+    const auto toAngle = rotaryStartAngle + sliderPos * (rotaryEndAngle - rotaryStartAngle);
+
+    const juce::Colour fill = slider.findColour (juce::Slider::rotarySliderFillColourId);
+
+    const knobImages& images = getKnobImages (diameter);
+    g.drawImageAt (images.backgroundImage, x, y);
+
+    if (!isSmall) // marker dot
+    {
+        g.saveState();
```

Comment:
Please use `juce::Graphics::ScopedSaveState save (g);` instead

Reply:
Same thing here, code would be:
```
    if (!isSmall) // marker dot
    {
        {
            juce::Graphics::ScopedSaveState save (g);
            auto dot = centre.getPointOnCircumference (radius * 0.825f, toAngle);
            auto dotRadius = radius * 0.072f;
            juce::Rectangle<float> dotBounds { dot.x - dotRadius, dot.y - dotRadius, dotRadius * 2.0f, dotRadius * 2.0f };

            fillEllipse (g, dotBounds, fill);

            clipToEllipse (g, dotBounds);

            g.setColour (fill.brighter (0.15f));
            g.drawEllipse (dotBounds.translated (0.0f, -1.0f), 1.5f);

            g.setColour (fill.darker (0.3f));
            g.drawEllipse (dotBounds.translated (0.0f, 0.5f), 1.5f);
        }

        g.setColour (blackA092);
        g.drawEllipse (dotBounds.expanded (0.5f), 1.0f);
    }
```